packages/headless/src/features/analytics/analytics-utils.ts

```diff
@@ -265,21 +265,101 @@ function makePreparableAnalyticsAction<
   return rootAction as PreparableAnalyticsAction<EventType, StateNeeded>;
 }
 
-export const makeAnalyticsAction = <
+export type AnalyticsActionOptions<
   EventType extends AnalyticsType,
   StateNeeded extends
     StateNeededBySearchAnalyticsProvider = StateNeededBySearchAnalyticsProvider,
->(
-  prefix: string,
-  analyticsType: EventType,
-  getBuilder: (
+> = LegacyAnalyticsOptions<EventType, StateNeeded> & NextLegacyOptions;
+
+export interface NextLegacyOptions {
+  todo?: never; //Remove when we have at least one property in this interface
+}
+export interface LegacyAnalyticsOptions<
+  EventType extends AnalyticsType,
+  StateNeeded extends
+    StateNeededBySearchAnalyticsProvider = StateNeededBySearchAnalyticsProvider,
+> {
+  prefix: string;
+  analyticsType: EventType;
+  __legacy__getBuilder: (
     client: CoveoSearchPageClient,
     state: StateNeeded
-  ) => Promise<EventBuilder | null> | null,
-  provider: (getState: () => StateNeeded) => SearchPageClientProvider = (
-    getState
-  ) => new SearchAnalyticsProvider(getState)
-): PreparableAnalyticsAction<WrappedAnalyticsType<EventType>, StateNeeded> => {
+  ) => Promise<EventBuilder | null> | null;
+  __legacy__provider?: (
+    getState: () => StateNeeded
+  ) => SearchPageClientProvider;
+}
+
+export function makeAnalyticsAction<
+  EventType extends AnalyticsType,
+  StateNeeded extends
+    StateNeededBySearchAnalyticsProvider = StateNeededBySearchAnalyticsProvider,
+>(
+  prefix: string,
+  analyticsType: LegacyAnalyticsOptions<
+    EventType,
+    StateNeeded
+  >['analyticsType'],
+  __legacy__getBuilder: LegacyAnalyticsOptions<
+    EventType,
+    StateNeeded
+  >['__legacy__getBuilder'],
+  __legacy__provider?: LegacyAnalyticsOptions<
+    EventType,
+    StateNeeded
+  >['__legacy__provider']
+): PreparableAnalyticsAction<WrappedAnalyticsType<EventType>, StateNeeded>;
+export function makeAnalyticsAction<
+  EventType extends AnalyticsType,
+  StateNeeded extends
+    StateNeededBySearchAnalyticsProvider = StateNeededBySearchAnalyticsProvider,
+>({
+  prefix,
+  analyticsType,
+  __legacy__getBuilder,
+  __legacy__provider,
+}: AnalyticsActionOptions<EventType, StateNeeded>): PreparableAnalyticsAction<
+  WrappedAnalyticsType<EventType>,
+  StateNeeded
+>;
+export function makeAnalyticsAction<
+  EventType extends AnalyticsType,
+  StateNeeded extends
+    StateNeededBySearchAnalyticsProvider = StateNeededBySearchAnalyticsProvider,
+>(
+  ...params:
+    | [
+        LegacyAnalyticsOptions<EventType, StateNeeded>['prefix'],
+        LegacyAnalyticsOptions<EventType, StateNeeded>['analyticsType'],
+        LegacyAnalyticsOptions<EventType, StateNeeded>['__legacy__getBuilder'],
+        LegacyAnalyticsOptions<EventType, StateNeeded>['__legacy__provider']?,
+      ]
+    | [AnalyticsActionOptions<EventType, StateNeeded>]
+): PreparableAnalyticsAction<WrappedAnalyticsType<EventType>, StateNeeded> {
+  return params.length === 1
```

Comment:
As you said, this does make for some ugly and hard to read code, which I'm honestly fine with since it's temporary (famous last word).

Could we add some small comment that explains why this looks like this/that this is temporary ?
So that if someone not aware of your changes stumbles upon that, there's less "wtf" ?

Should there be some UT to ensure things don't get broken for the old system during the transition ? Or I guess we can just rely on atomic end to end tests ?